source/frontend-packages.rst

```diff
@@ -0,0 +1,198 @@
+to upgrade both your project and its version of
+`Margarita <https://github.com/caktusgroup/margarita>`__ in order to
+use these frontend packaging techniques.
+
+Upgrading to Margarita 1.4.0 will require adding a few files to
+the project, which configure it as a NodeJS package for the
+purposes of defining dependencies and build instructions for the
+frontend components.
+
+First, you will need a Gulpfile, which is our frontend version of the
+Fabfile. Second, you'll need a ``package.json`` file defining the project
+for NPM, and defining the initial set of package dependencies.
+
+You can download and add these from this documentation:
+
+* `gulpfile.js <_static/files/gulpfile.js>`__
+
+* `package.json <_static/files/package.json>`__
+
+and add them to the top level of your project. You'll need to modify
+both with the correct project name.
+
+Finally, you can add these extra lines to your ``.gitignore``::
+
+    node_modules
+    */static/js/bundle.js
+    */static/css
+
+**TODO** Finish the upgrade section. It is incomplete.
+
+Using Frontend Packages
+=======================
+
+Our frontend packaging includes configuration to support the newest version of
+the Javascript language ES2015, which includes many updates including proper
+module support and an import statement. To use packages installed in your project
+you can import the default export from the project under a name of your choosing::
+
+    import $ from 'jquery';
+
+For some packages, you'll likely want to import more than one thing from it::
+
+    import { AutoBind, jQueryClass } from './helpers/jquery';
+
+Organizing Project Frontend Code
+--------------------------------
+
+Packages aren't just for the third party code you depend on. You'll
+benefit from organizing your project's own Javascript the same way. This setup begins
+with a top-level module in your project at ``static/js/index.js``. This module
+is only expected to be an initial point to import all your dependencies and do
+very basic initialization.
+
+A note for projects that use jQuery: because of the global nature of jQuery and
+the hoisting nature of ES2015 import statements, it can be very difficult to import
+jQuery in a way that makes it easily available to other scripts, such as plugins.
+We have found a double import method as follows works around these issues:
+
+.. code-block:: javascript
+
+    import $ from 'jquery'
+    /* ... additional imports ... */
+
+    window.jQuery = window.$ = require('jquery')
+
+The first, native import will ensure jQuery is loaded early as a dependency. The second
+version uses the non-native ``require()`` function, which is the NodeJS format for
+module import before ES2015. This version is just a regular line of code and so it will
+*not* be hoisted and can properly inject the window globals ``jQuery`` and ``$``.
+
+Creating And Using Modules
+--------------------------
+
+New functionality may require you organize it into its own module. Use your best
```

Comment:
Once we start hammering out a standard framework, maybe this section could get expanded (y'know, advice about the usual dir patterns followed on Flux+React projects and so on).